pkg/controller/controller_binding.go

```diff
@@ -671,218 +346,64 @@ func (c *controller) reconcileServiceBindingDelete(binding *v1beta1.ServiceBindi
 	if err != nil {
-		s := fmt.Sprintf(
+		msg := fmt.Sprintf(
 			`References a non-existent %s "%s/%s"`,
 			pretty.ServiceInstance, binding.Namespace, binding.Spec.ServiceInstanceRef.Name,
 		)
-		glog.Warningf(pcb.Messagef("%s (%s)", s, err))
-		c.recorder.Event(binding, corev1.EventTypeWarning, errorNonexistentServiceInstanceReason, s)
-		setServiceBindingCondition(
-			toUpdate,
-			v1beta1.ServiceBindingConditionReady,
-			v1beta1.ConditionFalse,
-			errorNonexistentServiceInstanceReason,
-			"The binding references an ServiceInstance that does not exist. "+s,
-		)
-		if _, err := c.updateServiceBindingStatus(toUpdate); err != nil {
-			return false, err
-
-		}
-		return false, err
+		readyCond := newServiceBindingReadyCondition(v1beta1.ConditionFalse, errorNonexistentServiceInstanceReason, msg)
+		return c.processServiceBindingOperationError(binding, readyCond)
 	}
 
 	if instance.Status.AsyncOpInProgress {
-		s := fmt.Sprintf(
+		msg := fmt.Sprintf(
 			`trying to unbind to %s "%s/%s" that has ongoing asynchronous operation`,
 			pretty.ServiceInstance, binding.Namespace, binding.Spec.ServiceInstanceRef.Name,
 		)
-		glog.Info(pcb.Message(s))
-		c.recorder.Event(binding, corev1.EventTypeWarning, errorWithOngoingAsyncOperation, s)
-		setServiceBindingCondition(
-			toUpdate,
-			v1beta1.ServiceBindingConditionReady,
-			v1beta1.ConditionFalse,
-			errorWithOngoingAsyncOperation,
-			errorWithOngoingAsyncOperationMessage,
-		)
-		if _, err := c.updateServiceBindingStatus(toUpdate); err != nil {
-			return false, err
-		}
-		return false, fmt.Errorf("Ongoing Asynchronous operation")
+		readyCond := newServiceBindingReadyCondition(v1beta1.ConditionFalse, errorWithOngoingAsyncOperation, msg)
+		return c.processServiceBindingOperationError(binding, readyCond)
 	}
 
 	if instance.Spec.ClusterServiceClassRef == nil || instance.Spec.ClusterServicePlanRef == nil {
 		// TODO(#1562): ultimately here we need to use logic similar to what is done to determine the plan ID for
 		// deprovisioning. We need to allow a ServiceBinding to be deleted, with an unbind request sent to the broker,
 		// even if the ServiceInstance has been changed to a non-existent plan.
-		return false, fmt.Errorf("ClusterServiceClass or ClusterServicePlan references for Instance have not been resolved yet")
+		return fmt.Errorf("ClusterServiceClass or ClusterServicePlan references for Instance have not been resolved yet")
 	}
 
 	serviceClass, servicePlan, brokerName, brokerClient, err := c.getClusterServiceClassPlanAndClusterServiceBrokerForServiceBinding(instance, binding)
 	if err != nil {
-		return false, err // retry later
-	}
-
-	unbindRequest := &osb.UnbindRequest{
-		BindingID:  binding.Spec.ExternalID,
-		InstanceID: instance.Spec.ExternalID,
-		ServiceID:  serviceClass.Spec.ExternalID,
-		PlanID:     servicePlan.Spec.ExternalID,
+		return c.handleServiceBindingReconciliationError(binding, err)
```

Comment:
For consistency, I'd like to see the other two places where `getClusterServiceClassPlanAndClusterServiceBrokerForServiceBinding` is called also call `handleServiceBindingReconciliationError` when an error is returned. I understand that right now `getClusterServiceClassPlanAndClusterServiceBrokerForServiceBinding` does not return an `operationError`, so the behavior is the same regardless. But I expect that we eventually we want to set the ready condition to false when there is an error in `getClusterServiceClassPlanAndClusterServiceBrokerForServiceBinding`.